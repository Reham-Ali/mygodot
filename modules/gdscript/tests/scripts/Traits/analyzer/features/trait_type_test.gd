trait Shearable:
    func shear() -> void:
        print("Shearable ok")

trait Milkable:
    func milk() -> void:
        print("Milkable ok")

class FluffyCow:
    uses Shearable, Milkable

class FluffyBull:
    uses Shearable

class Bird:
    pass

func test():
    Utils.check((FluffyCow is Shearable) == true)
    Utils.check((FluffyCow is Milkable) == true)

    var bull = FluffyBull.new()
    Utils.check((bull is Shearable) == true)
    Utils.check((bull is Milkable) == false)

    var my_animals : Array = []
    my_animals.append(FluffyCow.new())
    my_animals.append(FluffyBull.new())
    my_animals.append(Bird.new())
    var count = 1
    for animal in my_animals:
        print("Animal ", count)
        if animal is Shearable:
            animal.shear()
        if animal is Milkable:
            animal.milk()
        count += 1

    print("ok")
