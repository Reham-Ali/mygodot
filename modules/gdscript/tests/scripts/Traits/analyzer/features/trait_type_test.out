GDTEST_OK
>> WARNING
>> Line: 34
>> UNSAFE_METHOD_ACCESS
>> The method "shear()" is not present on the inferred type "Variant" (but may be present on a subtype).
>> WARNING
>> Line: 36
>> UNSAFE_METHOD_ACCESS
>> The method "milk()" is not present on the inferred type "Variant" (but may be present on a subtype).
Animal 1
Shearable ok
Milkable ok
Animal 2
Shearable ok
Animal 3
ok
